fix(modern): perform an exact check on `BPF_TRACE_RAW_TP` attach type

Andreagit97 opened the pull request:
**What type of PR is this?**

/kind bug

**Any specific area of the project related to this PR?**

/area libscap-engine-modern-bpf

**Does this PR require a change in the driver versions?**

No

**What this PR does / why we need it**:

This PR performs a more detailed check on the `BPF_TRACE_RAW_TP` attach type. Actual `libbpf_probe_bpf_prog_type(BPF_PROG_TYPE_TRACING, NULL)` checks for the `BPF_TRACE_FENTRY` attach type presence, while we need to check for the `BPF_TRACE_RAW_TP` one. If `BPF_TRACE_FENTRY` is defined we are sure `BPF_TRACE_RAW_TP` is defined as well, in all other cases, we need to search for it in the `vmlinux` file.

**Which issue(s) this PR fixes**:

See https://github.com/falcosecurity/falco/issues/2792

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
fix(modern): perform an exact check on `BPF_TRACE_RAW_TP` attach type
```

FedeDP commented on userspace/libpman/src/configuration.c:
```diff
@@ -165,15 +169,126 @@ int pman_get_required_buffers()
+	struct utsname buf;
+	char *data = NULL;
+	FILE *f = NULL;
+	uname(&buf);
```
I'd check the return code here.

Andreagit97 replied:
The reason why i didn't check it is that the first path to check is `"/sys/kernel/btf/vmlinux"`. Since it doesn't need the kernel release version even if `uname` fails probably we don't want to do anything and check at least for the first path  `"/sys/kernel/btf/vmlinux"`. In the worst case, when `uname` fails, we are just trying to read wrong paths, but `faccessat` should stop us. I think this is the reason why also in `libbpf` this function is not checked, see here https://elixir.bootlin.com/linux/latest/source/tools/lib/bpf/btf.c#L4786

FedeDP replied:
Makes sense! Thanks!

Molter73 replied:
How about moving the inner part of the loop to a separate function, checking `/sys/kernel/btf/vmlinux` before calling `uname` and skipping the rest of the paths if `uname` fails?

Andreagit97 replied:
yeah it could be an idea, i'm not even sure if we want to keep the other paths, in container environment they would require extra volume mounts of `/boot` or `lib/modules` :/

Molter73 replied:
Well, we mount `/usr/lib` to `/host/usr/lib` in our use case, it'd bother me a bit if this started failing on some platform, I'd at least try to access them through `scap_get_host_root` as a fallback.

Andreagit97 replied:
Uhm if you mount `/usr/lib` under `/host/usr/lib` it means that libbpf won't be able to perform these checks in your container https://elixir.bootlin.com/linux/latest/source/tools/lib/bpf/btf.c#L4786, is it true? or are you using some other workaround? This is in some way related to this https://github.com/falcosecurity/libs/pull/1404#discussion_r1355566414

Andreagit97 replied:
Solved here https://github.com/falcosecurity/libs/pull/1404#discussion_r1363566226

FedeDP reviewed (dismissed):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1404#" title="Author self-approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1404#pullrequestreview-1671501873" title="Approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
LGTM label has been added.  <details>Git tree hash: 1b6caf1ce9f3dd308f1b7988ad6a69ceadda9f07</details>

Andreagit97 commented:
/hold

Molter73 commented on userspace/libpman/src/configuration.c:
```diff
@@ -165,15 +169,126 @@ int pman_get_required_buffers()
+				z++;
+				if(z == sizeof(byte_array) / sizeof(*byte_array))
+				{
+					fclose(f);
+					free(data);
+					return true;
```
We can use an auxiliar `bool` variable like `res`, set it to false on instantiation, check it on the loop condition and use it to make this a bit more consistent with the rest of the code.
```suggestion
					res = true;
					goto loop_again;
```
We could also rename `loop_again` to something that is a bit more consistent, if you split this to its own function it could just be something like `end`

Andreagit97 replied:
done :)

Molter73 commented on userspace/libpman/src/configuration.c:
```diff
@@ -165,15 +169,126 @@ int pman_get_required_buffers()
+	// https://elixir.bootlin.com/linux/latest/source/tools/lib/bpf/btf.c#L4767
+	const char *locations[] = {
+		/* try canonical vmlinux BTF through sysfs first */
+		"/sys/kernel/btf/vmlinux",
```
Do we have access to `scap_get_host_root` from pman? It'd be great if this didn't require precision mounts from a user.

Andreagit97 replied:
Uhm I need to check it. The point is that even if here we use HOST_ROOT under the hood libbpf does exactly this same check with these paths to find btf on the system... So in this function we will search under /HOST_ROOT/boot/... while in libbpf case we will search under /boot/... Probably is better to use the same path between our code and libbpf, WDYT?

incertum replied:
`/sys/kernel/btf/vmlinux` should succeed in most cases also the container one as you highlighted Andrea. But am gonna +1 @Molter73 comment here. We know that when deploying in a container the other paths definitely won't exist. So I would add `scap_get_host_root` to be consistent with scap as well. That way we would check valid paths if we even get to them in the loop.

See also our guidance on host mounts needed https://falco.org/docs/install-operate/running/#docker-least-privileged

Andreagit97 replied:
I got your point, what is unclear to me is how `scap_get_host_root` could solve our issues here... 
* Let's say I add `scap_get_host_root` to this method and users will run a container with `-v /boot:/host/boot` so with `HOST_ROOT=/host`. 
* Imagine the case in which the vmlinux is not found under `/sys/kernel/btf/vmlinux` we will search it under  `/host/boot/vmlinux-%1$s` and we will find it.
* Now our check passes so we go on and we try to load the modern bpf with libbpf and we search for the btf file here https://elixir.bootlin.com/linux/latest/source/tools/lib/bpf/btf.c#L4773.
*  we won't fine the vmlinux under `/sys/kernel/btf/vmlinux` as before, we will search it under `/boot/vmlinux-%1$s` but we won't find anything since `/boot` is mounted under `/host/boot`

Molter73 replied:
I'm a bit torn on this one, I guess since `libbpf` limits us then we'll just have to keep it as is, `/sys/kernel/btf/vmlinux` should be enough on most cases anyways so, meh. WDYT @incertum?

incertum replied:
Yeah can't have nice things here @Molter73! As everyone is aware `/sys/kernel/btf/vmlinux` should be valid in probably almost all cases, so maybe we are over discussing something not relevant here. Let's merge then!

Andreagit97 replied:
Agree let's keep the paths as they are in libbpf

poiana commented:
New changes are detected. LGTM label has been removed.

Andreagit97 commented:
I have addressed the suggestions (thank you @Molter73). The actual solution checks for `/sys/kernel/btf/vmlinux` as a first fallback (IMO 99% of the cases), in the unlucky case in which the vmlinux is not found the user could be asked to mount `/boot`, `/lib` and `/usr` inside the container, the same requirement of `libbpf` (https://elixir.bootlin.com/linux/latest/source/tools/lib/bpf/btf.c#L4786). Let me know if you have other suggestions/ideas :)

FedeDP reviewed (dismissed):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 83335278ad86413233f815bc588b02043156afc4</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1404#" title="Author self-approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1404#pullrequestreview-1674081470" title="Approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 commented:
/unhold

Andreagit97 commented:
Since we didn't find an agreement in time I would move it to the next release

/milestone 0.14.0

Molter73 commented on userspace/libpman/src/configuration.c:
```diff
@@ -165,15 +169,138 @@ int pman_get_required_buffers()
 	return g_state.n_required_buffers;
 }
 
+static char byte_array[] = "BPF_TRACE_RAW_TP";
+
+bool check_location(const char* path)
+{
```
Since this is only used in `check_location` and `byte_array` is a pretty generic name, I would move it inside the function. By making it static it will not be initialized on every call anyways, so we should be OK (also, `const` because why not?). You could also rename it to something like `bpf_trace_raw`, but I'm not as concerned with naming.
```suggestion
bool check_location(const char* path)
{
	static const char byte_array[] = "BPF_TRACE_RAW_TP";
```

Andreagit97 replied:
Done, thank you!

Molter73 reviewed (commented):
Thanks for refactoring @Andreagit97! Code looks much cleaner this way IMO! Just left one more comment and would like to resolve the comment on pathing before merging, so waiting on @incertum for that one.

Molter73 reviewed (dismissed):
/lgtm

FedeDP commented:
@Andreagit97 small conflict and we are good to go!

poiana commented:
New changes are detected. LGTM label has been removed.

Molter73 reviewed (approved):
/lgtm

FedeDP reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 751cb33f7112c0c139664d637e88d390490bfd97</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1404#" title="Author self-approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1404#pullrequestreview-1684586686" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1404#pullrequestreview-1684586623" title="LGTM">Molter73</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP,Molter73]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
